src/header.rs

```diff
@@ -183,15 +186,18 @@ fn extract_full_array_shape(mut dtype: &DType) -> (Vec<u64>, &DType) {
 pub(crate) fn convert_value_to_shape(field: &Value) -> io::Result<Vec<u64>> {
     if let Value::List(ref lengths) = *field {
         lengths.iter().map(convert_value_to_shape_integer).collect()
+    } else if let Value::Tuple(ref lengths) = *field {
+        lengths.iter().map(convert_value_to_shape_integer).collect()
     } else {
         Err(invalid_data("shape must be list or tuple"))
     }
 }
 
 pub fn convert_value_to_shape_integer(number: &Value) -> io::Result<u64> {
-    if let Value::Integer(number) = *number {
-        if number >= 0 {
-            Ok(number as u64)
+    if let Value::Integer(number) = number {
+        let parts = number.to_u64_digits().1;
+        if !parts.is_empty() && parts[0] > 0 {
```

Comment:
Here is this logic again.  Especially given the other comments on it above, I would suggest factoring it out; perhaps rename this function to `convert_value_to_nonnegative_integer`, and then implement `convert_value_to_positive_integer` in terms of it.